cmd/syncComparator.go

```
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+//       both data+metadata, if dataChanged is not true then it should check metadataChanged and if that is true, sync only metadata, else sync nothing.
+//
+func (f *syncDestinationComparator) HasFileChangedSinceLastSyncUsingLocalChecks(so StoredObject, filePath string) (dataChange bool, metadataChange bool) {
+	// CFDMode==TargetCompare always treats target directories as “changed”, so we should never be called for that
+	if f.cfdMode == common.CFDModeFlags.TargetCompare() {
+		panic("We should not be called for CFDMode==TargetCompare")
+	}
+
+	// Changed file detection using Ctime and Mtime.
+	if f.cfdMode == common.CFDModeFlags.CtimeMtime() {
+		// File Mtime changed, which means data changed and it cause metadata change.
+		if so.lastModifiedTime.After(f.lastSyncTime) {
+			return true, true
+		} else if so.lastChangeTime.After(f.lastSyncTime) {
+			// File Ctime changed only, only meta data changed.
+			return false, true
+		}
+		// File not changed at all.
+		return false, false
+	} else if f.cfdMode == common.CFDModeFlags.Ctime() {
+		// Changed file detection using Ctime only.
+
+		// File changed since lastSync time. CFDMode is Ctime, so we can't rely on mtime as it can be modified by any other tool.
+		if so.lastChangeTime.After(f.lastSyncTime) {
+			// If MetaDataSync Flag is false we don't need to check for data or metadata change. We can return true in that case.
+			if !f.metaDataOnlySync {
+				return true, true
+			} else {
+				panic("We should not reach here, for CFDMODE==Ctime and metaDataOnlySync==true. It should be taken care with FinalizeAll flag.")
+			}
+		} else {
+			// File Ctime not changed, means no data or metadata changed.
+			return false, false
+		}
+	} else {
+		// This is the case when neither CtimeMtime or Ctime cfdMode set. So its target traverse and if we reach here
+		// means these entries for new files.
+		return true, true
```

we already checked and panic'ed for TargetCompare above.
Here we should add the same panic that says unsupported CFDmode and prints it